### PR TITLE
applet.sensor.qmc5883p: initial implementation

### PR DESCRIPTION
**Pull request by i-infra**
Initial implementation for the qmc5883p magnetometer sensor - i2c @ 0x2C. 

This code is tested working with the GY-271 breakout ie) https://www.aliexpress.us/item/2251832421287528.html. 

Shoutout @adafruit (referenced their libs)

Datasheet for part @ https://www.qstcorp.com/upload/pdf/202512/2C939E5AA0704285BC3BE71132B8629B.pdf 

PS: sending sm love to the glasgow team <3

### CODE REVIEW

File: software/glasgow/applet/sensor/qmc5883p/__init__.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+    8: DSR_8,
+}
+
+# Range names for user interface
+range_names = {
+    30: RANGE_30G,
+    12: RANGE_12G,
+    8: RANGE_8G,
+    2: RANGE_2G,
+}
```
Comment (whitequark) [inline software/glasgow/applet/sensor/qmc5883p/__init__.py]: Could you make these enums, please?

I understand that you've copied an implementation that is already in tree and using this particular pattern. However, the pattern is not very good (my bad) and I'd like to limit the amount of applets that need to be upgraded.

File: software/glasgow/applet/sensor/qmc5883p/__init__.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
+# Ref: QMC5883P Datasheet - Triple Axis Magnetometer
+# Accession: G00XXX
+
+import logging
+import asyncio
+import struct
+
+from ....support.data_logger import DataLogger
+from ... import *
+from ...interface.i2c_initiator import I2CInitiatorApplet
```
Comment (whitequark) [inline software/glasgow/applet/sensor/qmc5883p/__init__.py]: Please use absolute imports. (Same note applies as below.)

**Review by whitequark — CHANGES_REQUESTED** [review]
At a glance this looks great! The only issues I can see without closely looking through the functionality is ones that were already a problem in the codebase before you started working on your applet.

Please do the following in addition to inline comments:
- Add yourself to CODEOWNERS for this applet
- Add a documentation page at least for the CLI
- Add types to the `Interface` class (if this is something you would normally use in your code)
- Document the `Interface` class and add an API documentation section to the documentation page (I will merge the applet without this but it is preferred to have API docs for new applets like this one)

File: software/glasgow/applet/sensor/qmc5883p/__init__.py [hunk 3]
```diff
@@ -1,13 +1,61 @@
...
+Basic usage::
+
+    iface = await applet.run(device, args)
+    await iface.identify()
+    await iface.set_mode(OperatingMode.CONTINUOUS)
+    await iface.set_range(FieldRange.RANGE_8G)
+    x, y, z = await iface.get_magnetic()
+
+See ``QMC5883PInterface`` class documentation for available methods.
+"""
```
Comment (whitequark) [inline software/glasgow/applet/sensor/qmc5883p/__init__.py]: This should be in the `.rst` file under `docs/manual/applets/`. You shouldn't need to write any of this explicitly, just call into the argparse autodoc and the normal autodoc.

File: software/glasgow/applet/sensor/qmc5883p/__init__.py [hunk 4]
```diff
@@ -114,37 +181,91 @@ class QMC5883PError(GlasgowAppletError):
...
+    set_setreset_mode(setreset: SetResetMode) -> None
+        Set set/reset mode for eliminating offset.
+    data_ready() -> bool
+        Check if new measurement data is available.
+    overflow() -> bool
+        Check if sensor has overflowed.
+    get_magnetic_raw() -> tuple[int, int, int]
+        Read raw 16-bit signed magnetic field values (x, y, z).
+    get_magnetic() -> tuple[float, float, float]
+        Read magnetic field in Gauss (x, y, z).
```
Comment (whitequark) [inline software/glasgow/applet/sensor/qmc5883p/__init__.py]: These methods should be documented inline in the method docstring. (I think you may be using a different documentation style?)

Take a look at an existing applet, such as `i2c_controller`, to see an example of the proper style.

**Review by whitequark — CHANGES_REQUESTED** [review]
I did a rough pass (not a final one), looks better. By the way, I'm not sure why but "requesting review" from me on GitHub doesn't seem like it sends me an email; leaving a comment is the best way to get my attention.

**Comment by i-infra** [comment]
@whitequark I think this is good for re-review!

**Comment by whitequark** [comment]
Thanks, will take a look once I'm done with some urgent bureaucracy.

File: software/glasgow/applet/sensor/qmc5883p/__init__.py [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+_CHIPID = 0x00
+_XOUT_LSB = 0x01
+_XOUT_MSB = 0x02
+_YOUT_LSB = 0x03
+_YOUT_MSB = 0x04
+_ZOUT_LSB = 0x05
+_ZOUT_MSB = 0x06
+_STATUS = 0x09
+_CONTROL1 = 0x0A
+_CONTROL2 = 0x0B
```
Comment (whitequark) [inline software/glasgow/applet/sensor/qmc5883p/__init__.py]: Could you also use an enum here? 

Also, generally I avoid making things such as "register enumerations" `IntEnum`s unless there's a specific reason to do so; accidentally doing something like `_XOUT_LSB + 1` (in a more complex code path) and getting a nonsense result hurts a lot when debugging.

File: software/glasgow/applet/sensor/qmc5883p/__init__.py [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+                raise QMC5883PError(
+                    f"Invalid mode: {mode} (choose from: {', '.join(mode_names.keys())})"
+                )
+            mode = mode_names[mode]
+        elif mode not in mode_names.values():
+            raise QMC5883PError(f"Invalid mode: {mode}")
+
+        ctrl1 = await self._read_reg8u(_CONTROL1)
+        ctrl1 = (ctrl1 & ~0x03) | mode
+        await self._write_reg8u(_CONTROL1, ctrl1)
```
Comment (whitequark) [inline software/glasgow/applet/sensor/qmc5883p/__init__.py]: Please look into using `bitstruct` to avoid having to write error-prone manual bitwise operation sequences.

File: software/glasgow/applet/sensor/qmc5883p/__init__.py [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+    FieldRange.RANGE_2G: 15000.0,
+}
+
+# Mode names for user interface
+mode_names = {
+    "suspend": OperatingMode.SUSPEND,
+    "normal": OperatingMode.NORMAL,
+    "single": OperatingMode.SINGLE,
+    "continuous": OperatingMode.CONTINUOUS,
+}
```
Comment (whitequark) [inline software/glasgow/applet/sensor/qmc5883p/__init__.py]: I see what you've done here, and it's not a terrible implementation, but I think it can be significantly improved, here's how.

Right now, your enumeration values are the bit patterns you write to the device. This means the enumeration itself is pretty useless as far as UI is concerned.

Here's my proposal: make `OperatingMode.Suspend`'s value to be `"suspend"`. This gives you:
- the ability to use it as-is in `choices=OperatingMode`
- the ability to cast it from a string in the `Interface` class methods using something like `mode = mode if isinstance(mode, OperatingMode) else OperatingMode(mode)` (maybe in a helper method but I'm fine with it open-coded)
- the ability to convert things back, like in `FieldRange`, by just calling `.value` on the result

The conversion to/from bit patterns can be done using something like:

```python
def to_device(self) -> int:
    match self:
        case self.OperatingMode.Suspend: return 0
        case self.OperatingMode.Normal: return 1
        case self.OperatingMode.Single: return 2
        case self.OperatingMode.Continuous: return 3
```

and you mostly don't need to add explicit error handling since `Enum` will do it for you already.

File: software/glasgow/applet/sensor/qmc5883p/__init__.py [hunk 8]
```diff
@@ -0,0 +1,598 @@
...
+            await self._i2c_iface.write(self._i2c_address, [addr])
+            result = await self._i2c_iface.read(self._i2c_address, size)
+        if result is None:
+            raise QMC5883PError(
+                f"QMC5883P did not acknowledge I2C read at address {self._i2c_address:#04x}"
+            )
+        return list(result)
+
+    async def write(self, addr: int, data: list[int]) -> None:
+        await self._i2c_iface.write(self._i2c_address, [addr, *data])
```
Comment (whitequark) [inline software/glasgow/applet/sensor/qmc5883p/__init__.py]: This sensor *only* has an I2C interface, right? This intermediate `I2CInterface` class should only be used when there's multiple options, like I2C/SPI; otherwise it's just extra complexity for no benefit.

File: software/glasgow/applet/sensor/qmc5883p/__init__.py [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+
+
+class SensorQMC5883PApplet(GlasgowAppletV2):
+    logger = logging.getLogger(__name__)
+    help = "measure magnetic field with QMC5883P triple-axis magnetometer"
+    description = """
+    Measure magnetic field using the QMC5883P triple-axis magnetometer sensor.
+
+    This applet only supports sensors connected via the I²C interface.
```
Comment (whitequark) [inline software/glasgow/applet/sensor/qmc5883p/__init__.py:475]: This is the only one, right?

**Review by whitequark — CHANGES_REQUESTED** [review]

**Comment by whitequark** [comment]
This should be mergeable after the last review is addressed.

**Comment by whitequark** [comment]
Ah sorry, missed two really important points I stated elsewhere:

- needs documentation (you should include both the CLI docs as well as the `Interface` class and every enumeration it references; check that there aren't dead links in the generated docs)
- needs tests (at a minimum an `assertBuilds` but I strongly prefer having record-replay tests for at least the basic functionality; should be easy here

File: docs/manual/src/applets/sensor/qmc5883p.rst [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+.. autoprogram:: glasgow.applet.sensor.qmc5883p:SensorQMC5883PApplet._get_argparser_for_sphinx("sensor-qmc5883p")
+    :prog: glasgow run sensor-qmc5883p
+
+
+API reference
+-------------
+
+.. module:: glasgow.applet.sensor.qmc5883p
+
+.. autoexception:: QMC5883PError
```
Comment (whitequark) [inline docs/manual/src/applets/sensor/qmc5883p.rst:18]: I think you're missing a lot of the newly added enumerations (the ones that now get surfaced in the public API).

File: software/glasgow/applet/sensor/qmc5883p/__init__.py (hunk 5 above)
Comment (whitequark) [reply software/glasgow/applet/sensor/qmc5883p/__init__.py]: I think you're missing the enum here, unless it's intentional.

```
class Register(enum.Enum):
    CHIPID = 0x00
    XOUT_LSB = 0x01
    ...
```

**Review by whitequark — CHANGES_REQUESTED** [review]
Should be mergeable after the latest round of changs and maybe some tests!

**Comment by i-infra** [comment]
Think that's handled @whitequark - thanks for patience ; had a few other projects take up a lot of time of late (ie, weather is nice in the USA and the household needs to stack 3+ "cords" of wood before winter & we can't afford to heat the house any other way)

**Review by whitequark — APPROVED** [review]
Looks good! If you rebase and squash the commits I'll hit merge.

**Comment by i-infra** [comment]
@whitequark rebased/squashed; should be ready for merge. 🙏

**Comment by whitequark** [comment]
Thanks!
